gusto/linear_solvers.py

```diff
@@ -221,6 +222,214 @@ def solve(self):
+        beta = dt*state.timestepping.alpha
+        cp = state.parameters.cp
+        mu = state.mu
+        Vu = state.spaces("HDiv")
+        Vu_broken = FunctionSpace(state.mesh, BrokenElement(Vu.ufl_element()))
+        Vtheta = state.spaces("HDiv_v")
+        Vrho = state.spaces("DG")
+
+        h_deg = state.horizontal_degree
+        v_deg = state.vertical_degree
+        Vtrace = FunctionSpace(state.mesh, "HDiv Trace", degree=(h_deg, v_deg))
+
+        # Split up the rhs vector (symbolically)
+        u_in, rho_in, theta_in = split(state.xrhs)
+
+        # Build the reduced function space for u,rho
+        M = MixedFunctionSpace((Vu_broken, Vrho))
+        w, phi = TestFunctions(M)
+        u, rho = TrialFunctions(M)
+
+        l0 = TrialFunction(Vtrace)
+        dl = TestFunction(Vtrace)
+
+        n = FacetNormal(state.mesh)
+
+        # Get background fields
+        thetabar = state.fields("thetabar")
+        rhobar = state.fields("rhobar")
+        pibar = exner(thetabar, rhobar, state)
+        pibar_rho = exner_rho(thetabar, rhobar, state)
+        pibar_theta = exner_theta(thetabar, rhobar, state)
+
+        # Analytical (approximate) elimination of theta
+        k = state.k             # Upward pointing unit vector
+        theta = -dot(k,u)*dot(k,grad(thetabar))*beta + theta_in
+
+        # Only include theta' (rather than pi') in the vertical
+        # component of the gradient
+
+        # the pi prime term (here, bars are for mean and no bars are
+        # for linear perturbations)
+
+        pi = pibar_theta*theta + pibar_rho*rho
+
+        # vertical projection
+        def V(u):
+            return k*inner(u,k)
+
+        # specify degree for some terms as estimated degree is too large
+        dxp = dx(degree=(self.quadrature_degree))
+        dS_vp = dS_v(degree=(self.quadrature_degree))
+        dS_hp = dS_h(degree=(self.quadrature_degree))
+        ds_vp = ds_v(degree=(self.quadrature_degree))
+        ds_tbp = ds_t(degree=(self.quadrature_degree)) + ds_b(degree=(self.quadrature_degree))
+
+        rhobar_tr = Function(Vtrace)
+        rbareqn = (l0('+') - avg(rhobar))*dl('+')*(dS_vp + dS_hp) + \
+                  (l0 - rhobar)*dl*ds_vp + \
+                  (l0 - rhobar)*dl*ds_tbp
+        rhobar_prob = LinearVariationalProblem(lhs(rbareqn),rhs(rbareqn),rhobar_tr)
+        self.rhobar_solver = LinearVariationalSolver(rhobar_prob,
+                                                     solver_parameters={'ksp_type':'preonly',
+                                                                        'pc_type':'bjacobi',
+                                                                        'pc_sub_type':'lu'})
+
+        Aeqn = (
+            inner(w, (u - u_in))*dx
+            - beta*cp*div(theta*V(w))*pibar*dxp
+            - beta*cp*div(thetabar*w)*pi*dxp
+            + (phi*(rho - rho_in) - beta*inner(grad(phi), u)*rhobar)*dx
+            + beta*inner(phi*u, n)*rhobar_tr*(dS_v + dS_h)
```

Comment:
A comment: the compile error you're experiencing is due to the form:
```
beta*inner(phi*u, n)*rhobar_tr*(dS_v + dS_h)
```
However, if you remove `rhobar_tr`, the tensor assembles without error. So the problem seems to be in `rhobar_tr`.

I can also verify that Slate can indeed assemble forms with trace coefficients in both extruded and non-extruded cases. I just verified this moments ago. I'm still looking into the issues and will update.